drop-in/src/main/java/com/adyen/checkout/dropin/ui/viewmodel/PaymentMethodsListDelegate.kt

```diff
@@ -0,0 +1,198 @@
+                }
+                PaymentMethodTypes.UNSUPPORTED_PAYMENT_METHODS.contains(type) -> {
+                    Logger.e(TAG, "PaymentMethod not yet supported - $type")
+                    checkPaymentMethodsMap[paymentMethod] = false
+                }
+                else -> {
+                    Logger.d(TAG, "No availability check required - $type")
+                    checkPaymentMethodsMap[paymentMethod] = true
+                }
+            }
+        }
+        checkIfListReady()
+    }
+
+    override fun onAvailabilityResult(isAvailable: Boolean, paymentMethod: PaymentMethod, config: Configuration?) {
+        Logger.d(TAG, "onAvailabilityResult - ${paymentMethod.type}: $isAvailable")
+        checkPaymentMethodsMap[paymentMethod] = isAvailable
+        checkIfListReady()
+    }
+
+    private fun checkIfListReady() {
+        if (paymentMethods.size == checkPaymentMethodsMap.size) {
+            populatePaymentMethods()
+        }
+    }
+
+    private fun populatePaymentMethods() {
+        val paymentMethods = mutableListOf<PaymentMethodListItem>().apply {
+            // gift cards
+            val giftCardsList: List<GiftCardPaymentMethodModel> =
+                order?.paymentMethods?.mapToGiftCardPaymentMethodModel().orEmpty()
+            if (giftCardsList.isNotEmpty()) {
+                add(PaymentMethodHeader(PaymentMethodHeader.TYPE_GIFT_CARD_HEADER))
+                addAll(giftCardsList)
+            }
+            // payment notes
+            order?.remainingAmount?.let { remainingAmount ->
+                val value = CurrencyUtils.formatAmount(remainingAmount, dropInConfiguration.shopperLocale)
+                add(
+                    PaymentMethodNote(application.getString(R.string.checkout_giftcard_pay_remaining_amount, value))
+                )
+            }
+            // stored payment methods
+            storedPaymentMethodsList = storedPaymentMethods.setupStoredPaymentMethods().toMutableList()
+            val hasStoredPaymentMethods = storedPaymentMethodsList.isNotEmpty()
+            if (hasStoredPaymentMethods) {
+                add(PaymentMethodHeader(PaymentMethodHeader.TYPE_STORED_HEADER))
+                addAll(storedPaymentMethodsList)
+            }
+            // payment methods
+            val paymentMethodsList: List<PaymentMethodModel> = paymentMethods.setupPaymentMethodsList()
+            if (paymentMethodsList.isNotEmpty()) {
+                val headerType = if (hasStoredPaymentMethods)
+                    PaymentMethodHeader.TYPE_REGULAR_HEADER_WITHOUT_STORED
+                else
+                    PaymentMethodHeader.TYPE_REGULAR_HEADER_WITH_STORED
+                add(PaymentMethodHeader(headerType))
+                addAll(paymentMethodsList)
+            }
+        }
+        _paymentMethodsFlow.tryEmit(paymentMethods)
+    }
+
+    internal fun removePaymentMethodWithId(id: String) {
+        if (this::storedPaymentMethodsList.isInitialized) {
```

Comment:
I'm not a big fan of this approach, it's the same as having a nullable list and checking whether it's null or not, except that now someone might call `storedPaymentMethodsList` directly somewhere else and cause a crash.
Maybe make the list nullable?